metals/src/main/scala/scala/meta/internal/metals/Diagnostics.scala

```diff
@@ -77,6 +77,16 @@ final class Diagnostics(
     compileTimer.remove(target)
   }
 
+  def onSyntaxError(path: AbsolutePath, diags: List[Diagnostic]): Unit = {
+    diags.headOption match {
+      case Some(diagnostic) if !path.isInReadonlyDirectory(workspace) =>
+        syntaxError(path) = diagnostic
+        publishDiagnostics(path)
```

Comment:
This confuses me a bit. If we have the list of diagnostics already, how come we check to see if it's nonEmpty, and if so, we do `publishDiagnostics(path)`, which then does a lookup to get the diagnostics for that path. But we already have the diagnostics here. Couldn't we just pass them?

Reply:
This might have made sense at some point, but might have lost it due to refactoring. I will try to simplify it!

Reply:
Ok, so the difference here is that we treat syntax errors differently to the errors coming from the build server. At each publish we combine both to make sure there are not duplicates. That's why we save the syntax error here, then run publishDiagnostics that will actually combine them both.

We could rework this a bit for sure as it is not clear at all, but I don't think it should be done in this PR as it's already complex. Maybe create a follow up issue to simplify diagnostics?

Reply:
Ahhh I totally didn't know that. Thanks for the explanation! As for the re-work, yea, add it to the long list.